### PR TITLE
feat: consolidate tractus-x HashiCorp Vault impl

### LINKED ISSUE #3254: Consolidate multiple implementations of `HashicorpVault`
# Feature Request

Currently there is one implementation of the HashiCorp Vault directly in the Connector repository, but there's also one in [Tractus-X](https://github.com/eclipse-tractusx/tractusx-edc/tree/main/edc-extensions/hashicorp-vault)

We should consolidate these two implementations and host them upstream in the Connector repo. For example, the downstream impl contains a health check, which the upstream impl lacks.

When doing that, we may want to relax our guidelines towards Test containers, since the Tractus-X impl heavily relies on them.

## Which Areas Would Be Affected?

control plane

## Why Is the Feature Desired?

Consolidate efforts, avoid code duplication. 

## Solution Proposal

- [ ] #3255 
- [ ] move code from Tractus-X into EDC, merging the implementations
- [ ] delete the downstream impl.

### PR DESCRIPTION
**Pull request by paullatzelsperger**
## What this PR changes/adds

This PR takes the developments of the [Tractus-X](https://github.com/eclipse-tractusx/tractusx-edc/tree/main/edc-extensions/hashicorp-vault) implementation of HashiCorp Vault and upstreams it to EDC.

## Why it does that

Consolidating efforts, avoid duplication.

## Further notes

- this is the first PR that actually utilizes test containers in EDC!
- no functional changes of the code base (some minor refactorings aside)
- once this PR is merged, and safely packaged in an official version, it will be removed from Tractus-X EDC

## Linked Issue(s)

Closes #3254 

_Please be sure to take a look at the [contributing guidelines](https://github.com/eclipse-edc/Connector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and our [etiquette for pull requests](https://github.com/eclipse-edc/Connector/blob/main/pr_etiquette.md)._

### CODE REVIEW

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java [hunk 1]
```diff
@@ -9,100 +9,281 @@
...
-import static org.eclipse.edc.vault.hashicorp.HashicorpVaultExtension.VAULT_TOKEN;
-import static org.eclipse.edc.vault.hashicorp.HashicorpVaultExtension.VAULT_URL;
+import static org.eclipse.edc.vault.hashicorp.HashicorpVaultClient.VAULT_DATA_ENTRY_NAME;
+import static org.eclipse.edc.vault.hashicorp.model.Constants.VAULT_TOKEN;
+import static org.eclipse.edc.vault.hashicorp.model.Constants.VAULT_URL;
+import static org.eclipse.edc.vault.hashicorp.util.X509CertificateTestUtil.convertToPem;
+import static org.eclipse.edc.vault.hashicorp.util.X509CertificateTestUtil.generateCertificate;
 
-@HashicorpVaultTest
+@ComponentTest
```
Comment (ndr-brt) [inline extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java:60]: maybe I missed this, so from now on all the edge tests with external services will become component tests?
I'm fine with it, but the "component tests" job already takes like 11 minutes, my fear is that adding other tests that involve test containers this time could raise significantly

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java [hunk 2]
```diff
@@ -9,100 +9,281 @@
...
 class HashicorpVaultIntegrationTest {
-    private static final String VAULT_TEST_URL = "http://127.0.0.1:8200";
-    private static final String VAULT_TEST_TOKEN = "test-token";
+    static final String DOCKER_IMAGE_NAME = "vault:1.9.6";
+    static final String VAULT_ENTRY_KEY = "testing";
+    static final String VAULT_ENTRY_VALUE = UUID.randomUUID().toString();
+    static final String TOKEN = UUID.randomUUID().toString();
+
+    @Container
+    @ClassRule
```
Comment (ndr-brt) [inline extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java:70]: If I remember correctly, the `@ClassRule` annotation was needed for junit4, for junit5 `@Container` should be enough

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java [hunk 3]
```diff
@@ -9,100 +9,281 @@
...
+    public static final VaultContainer<?> VAULTCONTAINER = new VaultContainer<>(DOCKER_IMAGE_NAME)
+            .withVaultToken(TOKEN)
+            .withSecretInVault("secret/" + VAULT_ENTRY_KEY, format("%s=%s", VAULT_DATA_ENTRY_NAME, VAULT_ENTRY_VALUE));
+
+    private final TestExtension testExtension = new TestExtension();
```
Comment (ndr-brt) [inline extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java:75]: is this needed? couldn't the services being registered directly on the `extension` in the `beforeEach` method?)

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java [hunk 4]
```diff
@@ -9,100 +9,281 @@
...
+        public CertificateResolver getCertificateResolver() {
+            return certificateResolver;
+        }
+
+        public Vault getVault() {
+            return vault;
+        }
+    }
+
+    private static class MyHealthCheckService implements HealthCheckService {
```
Comment (ndr-brt) [inline extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java:235]: why is this needed?

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultTest.java [hunk 5]
```diff
@@ -16,62 +16,107 @@
...
+    // mocks
+    private HashicorpVaultClient vaultClient;
     private HashicorpVault vault;
 
     @BeforeEach
     void setup() {
-        vaultHashicorpVaultClient = mock(HashicorpVaultClient.class);
-        var monitor = mock(Monitor.class);
-        vault = new HashicorpVault(vaultHashicorpVaultClient, monitor);
+        vaultClient = Mockito.mock(HashicorpVaultClient.class);
```
Comment (ndr-brt) [inline extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultTest.java:38]: static import, and I discovered that, when the type of the variable is defined, you don't have to pass the type to the `mock` method, it will understand it implicitly

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/PathUtilTest.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+                Arguments.of("/v1/secret/data/", "v1/secret/data"),
+                Arguments.of("v1/secret/data/", "v1/secret/data"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideStringsForTrimsPathsCorrect")
+    void trimsPathsCorrect(String path, String expected) {
+        final String result = PathUtil.trimLeadingOrEndingSlash(path);
+
+        Assertions.assertEquals(expected, result);
```
Comment (ndr-brt) [inline extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/PathUtilTest.java:39]: static import

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/PathUtilTest.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.stream.Stream;
+
+class PathUtilTest {
+
+    private static Stream<Arguments> provideStringsForTrimsPathsCorrect() {
```
Comment (ndr-brt) [inline extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/PathUtilTest.java:26]: I know that I'm getting pedantic about a nit, but I thing using the `ArgumentSource` trough a class extending `ArgumentProvider` instead of the `MethodSource` is way better.

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java (hunk 1 above)
Comment (paullatzelsperger) [reply extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java:60]: as far as I understand it, testcontainers are spinning up some sort of a proxy container, which then launches the actual Postgres, etc. So running subsequent tests should be pretty quick, at least I could see that on my maching.
I used the `@ComponentTest` annotation, because now `@Postgres....Test` should be reserved for those that actually use an "external" postgres DB, i.e. E2E tests.

If we see that this causes grievances down the line, it should be easy to change.

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/PathUtilTest.java (hunk 7 above)
Comment (paullatzelsperger) [reply extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/PathUtilTest.java:26]: I agree with you, it simply slipped through during the copy-paste marathon :)

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/PathUtilTest.java (hunk 6 above)
Comment (paullatzelsperger) [reply extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/PathUtilTest.java:39]: will convert to an AssertJ assertion

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultTest.java (hunk 5 above)
Comment (paullatzelsperger) [reply extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultTest.java:38]: I've noticed it too, especially useful when the static type is implicit, such as for fields.

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java (hunk 4 above)
Comment (paullatzelsperger) [reply extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java:235]: probably not at all. will remove it.

File: extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java (hunk 3 above)
Comment (paullatzelsperger) [reply extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultIntegrationTest.java:75]: the weird thing is, that this extension wasn't even used to _register_ the vault, but rather to _obtain_ it...

**Review by ndr-brt — APPROVED** [review]
